jax/api.py

```diff
@@ -1966,6 +1966,40 @@ def device_put(x, device: Optional[xc.Device] = None):
   return tree_map(lambda y: xla.device_put_p.bind(y, device=device), x)
 
 
+def device_put_sharded(x, devices: Sequence[xc.Device]):
+  """Shards the input to specified devices, returning ShardedDeviceArrays.
+
+  Args:
+    x: A sequence of arrays, scalars, or (nested) standard Python containers thereof
+    devices: A sequence of devices()
+
+  Returns:
+    A ShardedDeviceArray or (nested) Python container thereof containing a copy
+    of x sharded across the specified devices.
+  """
+  # TODO(jakevdp): provide a default for devices that considers both local devices and pods
+  assert len(x) == len(devices), f"len(x) = {len(x)} must equal len(devices) = {len(devices)}."
+  def _device_put_sharded(*xs) -> pxla.ShardedDeviceArray:
+    aval = ShapedArray((len(xs),) + np.shape(xs[0]), dtypes.result_type(xs[0]))
+    buffers = [xla.device_put(x, d) for x, d in zip(xs, devices)]
+    return pxla.ShardedDeviceArray(aval, buffers)
+  return tree_multimap(_device_put_sharded, *x)
+
+
+def device_put_replicated(x, devices: Sequence[xc.Device]):
+  """Replicates the input across specified devices, returning ShardedDeviceArrays.
```

Comment:
I would add "broadcasted over an extra logical axis" or something to that effect (and/or add an API to create SDAs that are replicated _without_ that extra axis)

Reply:
If I understand your comment correctly, I think the version that replicates without the extra axis is `device_put_sharded`.

Reply:
Added the additional language as well.

Reply:
Hmm. By "replicates without the extra axis", I mean "creates an SDA that uses `replication_factors` to represent an array replicated over devices" (the aval of the SDA result is the same as the aval of the DA input).

Reply:
Maybe the issue is the terminoligy I'm using. I want this function to be something that takes an input and returns a sharded device array containing the input stacked over the number of devices, because empirically this is a common use-case in downstream libraries. Is there a better name for this than "replicate"?

Reply:
There are two separate questions: what is the function doing logically, and what is it doing in terms of devices/buffers:

- Logically, you can imagine functions that concatenate or stack a list of inputs, and functions that represent an identity or a broadcast of a single input.
- Device-wise, we want all of these to have the resulting array reside across multiple devices, but there are a few ways to do this: either an identical copy of the whole logical array on every device, or a separate piece of the logical array on each device.

Reply:
In MLPerf code, we had a `replicate` function with a `broadcast: bool` kwarg. When `broadcast=False`, it represented an identity function that results in an output SDA with the same aval as the input, but contains copies of the whole array on every device and uses `replication_factors` to encode this. When `broadcast=True`, it represented a broadcast function that results in an output SDA with an extra logical axis, containing copies of the whole _original_ array on every device (but now they're slices of the new logical array), and for which `replication_factors` is empty.

Reply:
OK, maybe more directly to the point: I want the functions to do what I wrote in the code here, because this is the functionality that downstream libraries need. What should the functions be called?

Reply:
Sorry for being a little obnoxious: it's the functionality that downstream libraries currently have, but not necessarily what they need/should have in the long run. A "replicate" function that's logically a broadcast is only needed while `pmap` doesn't fully support `(in/out)_axes=None`, and becomes tech debt as soon as that support lands (since the broadcast doesn't do anything for users, and e.g. makes this API useless with `soft_pmap`).

So I'd suggest using a name (or kwarg) with "broadcast" in it for this function. Maybe `api.device_put_broadcasted`? Or `api.device_put_replicated(broadcast=True)`? For the other one, `api.device_put_sharded` sounds fine.

Reply:
OK, I see. Maybe better then would be to remove the `device_put_replicated(x, devices)` function altogether, since it's functionally equivalent to `device_put_sharded([x] * len(devices), devices)` (though there's a bit more validation overhead this way).

Reply:
Alright, all that's left is the `device_put_sharded` function, which has clearer semantics.